tests/test_flow/test_logging.py

```diff
@@ -0,0 +1,90 @@
+import pytest
+
+import logging
+
+
+class LogChecker:
+    def __init__(self, caplog):
+        self._caplog = caplog
+
+    def expect(self, *expected_messages):
+        actual_messages = [record.getMessage() for record in self._caplog.records]
+        assert set(actual_messages) == set(expected_messages)
+        self._caplog.clear()
+
+
+@pytest.fixture(scope="function")
+def log_checker(caplog):
+    caplog.set_level(logging.INFO)
+    return LogChecker(caplog)
+
+
+def test(builder, log_checker):
+    """
+    Test the details of the log messages we emit. Since these messages are currently the
+    best way to get visibility into what Bionic is doing, we have much more detailed
+    tests than we'd normally want for logging. This means we'll have to tweak these
+    tests as we update the format or implementation details of our logging.
+
+    At some point we should introduce a separate system for user-facing
+    progress reporting instead of using logs.
+    """
+
+    builder.assign("x", 1)
+
+    @builder
+    def x_plus_one(x):
+        return x + 1
+
+    @builder
+    def x_plus_two(x_plus_one):
+        return x_plus_one + 1
+
+    flow = builder.build()
+    assert flow.get("x_plus_one") == 2
+
+    log_checker.expect(
+        "Accessed   x(x=1) from definition",
+        "Computing  x_plus_one(x=1) ...",
+        "Computed   x_plus_one(x=1)",
+    )
+
+    assert flow.get("x_plus_two") == 3
+    log_checker.expect(
+        # It would be reasonable to log the following line, but our
```

Comment:
I don't think this is 100% true. In-memory cache is only populated for a completed task. We do log in-memory cache access in `_compute_result_group_for_dnode`. So I think if `x_plus_two` depended on `x_plus_one` and `one` and `one` wasn't complete, we would have logged `x_plus_one` cache access if though the task was complete from the previous get.

This should be a pretty easy change, you just need to add an in-memory access log in `_get_results_for_complete_task_state`.

Reply:
You're right -- thanks!